h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -489,44 +505,82 @@ public MP nextModelParameters(Model previousModel) {
-              // ToDo: model seed setting will be different for parallel model building.
-              // ToDo: This implementation only works for sequential model building.
-              if (_set_model_seed_from_search_seed) {
-                // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
-                params._seed = _search_criteria.seed() + (model_number++);
-              }
+             
+            if (_filterFunctions == null || hyperParamsAreNotSkipped(hypers, _filterFunctions)) {
+
+              _numberOfNonSkippedPermutations++;
+              // Get clone of parameters
+              MP commonModelParams = (MP) _params.clone();
+              // Fill model parameters
+              MP params = getModelParams(commonModelParams, hypers);
+
+              // add max_runtime_secs in search criteria into params if applicable
+              if (_search_criteria != null && _search_criteria.strategy() == HyperSpaceSearchCriteria.Strategy.RandomDiscrete) {
+                // ToDo: model seed setting will be different for parallel model building.
+                // ToDo: This implementation only works for sequential model building.
+                if (_set_model_seed_from_search_seed) {
+                  // set model seed = search_criteria.seed+(0, 1, 2,..., model number)
+                  params._seed = _search_criteria.seed() + (model_number++);
+                }
 
-              // set max_runtime_secs
-              double timeleft = this.time_remaining_secs();
-              if (timeleft > 0)  {
-                if (params._max_runtime_secs > 0) {
-                  params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
-                } else {
-                  params._max_runtime_secs = timeleft;
+                // set max_runtime_secs
+                double timeleft = this.time_remaining_secs();
+                if (timeleft > 0) {
+                  if (params._max_runtime_secs > 0) {
+                    params._max_runtime_secs = min(params._max_runtime_secs, timeleft);
+                  } else {
+                    params._max_runtime_secs = timeleft;
+                  }
                 }
               }
+              return params;
+            } else {
+              
+              int nextPermutationNum = _currentPermutationNum + 1;
+              if(hasNext(previousModel) && (max_models() == 0 || max_models() > 0 && nextPermutationNum < max_models())) {
+                return nextModelParameters(previousModel);
+              } else {
+                return null;
+              }
             }
-            return params;
           } else {
             throw new NoSuchElementException("No more elements to explore in hyper-space!");
           }
         }
+        
+        private boolean hyperParamsAreNotSkipped(Object[] hypers, ArrayList<Function<Map<String, Object>, Boolean>> filterFunctions) {
+          boolean gridItemShouldBeSkipped = false;
```

Comment:
and sometimes, shorter is clearer: `skip` works fine.
We have a param named `hypers`, no need to introduce a new concept like `gridItem`, it makes it harder to understand what's happening..

Reply:
suggesting sth like:
```
boolean skip = false;
Map<String, Object> hypersAsMap = convertToMap(hypers, _hyperParamNames);

for(Function<Map<String, Object>, Boolean> fun : filterFunctions) {
      skip = !fun.apply(hypersAsMap);
      if (skip) break;
}
return skip;
```
not sure the additional decrement logic is necessary either (see comment below).
Although I have a doubt as I see that you're re-applying all the `KeepOnlyFirstMatchFilterFunction`, even those that were not tried in the first for-loop (due to the break as soon as we know we will skip).

Reply:
@seb-h2o 
> and sometimes (often?), shorter is clearer

in general I would not say it is true. To make something more clear we need to make sure that there is no ambiguities and there are no redundant words ( and it is not always shorter). But of course it is better to be concise.

> We have a param named hypers, no need to introduce a new concept like gridItem

I don't like `hypers` tbh. It is more generic than `gridItem`. Actually any sub set of hyper parameters could be called `hypers` as it just implies that we are dealing with multiple hyper parameters. `gridItem` implies that it is a **complete** set of hyper parameters. Btw we already have another naming for `hypers` - `permutation` ( see HyperSpaceWalker.java), which I think is closer to `gridItem`.   Thanks for bringing this up... I will think about options. Maybe we just need to unify naming.

Reply:
@seb-h2o `_hyperParams` (L327 `HyperSpaceWalker.java` ) for example looks like `hypers` but it refers to all permutations( grid), not just to one permutation. So there is a problem with naming, but it was there before.

Reply:
@seb-h2o tend to agree that without negation in naming and just `!` in front of method's call it looks better. Thanks!

Reply:
> Maybe we just need to unify naming.

that's my point, pick one, and stick to it.